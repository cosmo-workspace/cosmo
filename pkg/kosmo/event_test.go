package kosmo

import (
	"bytes"
	"context"
	"fmt"
	"os"
	"reflect"
	"slices"
	"strings"
	"testing"
	"time"

	. "github.com/cosmo-workspace/cosmo/pkg/snap"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"

	"github.com/google/go-cmp/cmp"
	corev1 "k8s.io/api/core/v1"
	eventsv1 "k8s.io/api/events/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/cli-runtime/pkg/printers"
	"k8s.io/client-go/tools/record"

	cosmov1alpha1 "github.com/cosmo-workspace/cosmo/api/v1alpha1"
	"github.com/cosmo-workspace/cosmo/pkg/kubeutil"
	"github.com/cosmo-workspace/cosmo/pkg/template"
)

var _ = Describe("event", func() {
	Context("when list user events", func() {
		It("should return a array of the events in user namespace and the user annotated events in default namespace", func() {
			ctx := context.Background()
			err := k8sClient.Create(ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "cosmo-user-xxx"}})
			Expect(err).ShouldNot(HaveOccurred())

			in, err := os.ReadFile("./test/events.yaml")
			Expect(err).ShouldNot(HaveOccurred())

			unst, err := template.NewRawYAMLBuilder(string(in)).Build()
			Expect(err).ShouldNot(HaveOccurred())

			for _, event := range unst {
				RemoveDynamicFields(&event)
				_, err := kubeutil.Apply(ctx, k8sClient, &event, "test", false, false)
				Expect(err).ShouldNot(HaveOccurred(), "failed to create event", event)
			}

			k := NewClient(k8sClient)
			events, err := k.ListEvents(ctx, "cosmo-user-xxx")
			Expect(err).ShouldNot(HaveOccurred())

			data := make([][]string, 0, len(events)+1)
			data = append(data, []string{"Event Time", "Type", "Reason", "Regarding", "Reporting Controller", "Note"})
			for _, v := range events {
				data = append(data, []string{eventTime(v).GoString(), v.Type, v.Reason, v.Regarding.Kind + "/" + v.Regarding.Name, v.ReportingController, v.Note})
			}

			out := bytes.Buffer{}
			w := printers.GetNewTabWriter(&out)

			for _, v := range data {
				fmt.Fprintf(w, "%s\n", strings.Join(v, "\t"))
			}

			w.Flush()
			Expect(out.String()).To(MatchSnapShot())
		})
	})
	Context("Eventf", func() {
		It("emit events", func() {
			eventCount := 4

			rec := record.NewFakeRecorder(eventCount)
			inst := cosmov1alpha1.Instance{
				ObjectMeta: metav1.ObjectMeta{
					Name:      "inst1",
					Namespace: "cosmo-user-tom",
				},
			}
			InstanceEventf(rec, &inst, corev1.EventTypeNormal, "TEST1", "Instance: key1=%s", "value1")

			cinst := cosmov1alpha1.ClusterInstance{
				ObjectMeta: metav1.ObjectMeta{
					Name: "cinst1",
				},
			}
			InstanceEventf(rec, &cinst, corev1.EventTypeNormal, "TEST2", "ClusterInstance: key1=%s", "value1")

			cinst2 := cosmov1alpha1.ClusterInstance{
				ObjectMeta: metav1.ObjectMeta{
					Name: "cinst2",
					OwnerReferences: []metav1.OwnerReference{
						{
							APIVersion: "cosmo.workspace/v1alpha1",
							Kind:       "User",
							Name:       "user1",
						},
					},
				},
			}
			InstanceEventf(rec, &cinst2, corev1.EventTypeWarning, "TEST3", "Addon: key1=%s", "value1")

			user := cosmov1alpha1.User{
				ObjectMeta: metav1.ObjectMeta{
					Name: "user2",
				},
			}
			UserEventf(rec, &user, corev1.EventTypeWarning, "TEST4", "User: key1=%s", "value1")

			i := 0
		L:
			for {
				select {
				case msg, ok := <-rec.Events:
					if ok {
						i++
						Expect(msg).To(MatchSnapShot())
					} else {
						Fail("CLOSED")
					}
				case <-time.After(time.Second * 10):
					Fail("TIMEOUT")
				default:
					break L
				}
			}
			Expect(i).To(Equal(eventCount))
		})
	})
})

func Test_sorted(t *testing.T) {
	type args struct {
		events []eventsv1.Event
	}
	tests := []struct {
		name string
		args args
		want []eventsv1.Event
	}{
		{
			name: "same DeprecatedFirstTimestamp; diff DeprecatedLastTimestamp",
			args: args{
				events: []eventsv1.Event{
					{
						DeprecatedFirstTimestamp: metav1.Date(2024, 5, 26, 15, 43, 16, 0, time.UTC),
						DeprecatedLastTimestamp:  metav1.Date(2024, 5, 26, 15, 44, 35, 0, time.UTC),
						DeprecatedSource: corev1.EventSource{
							Component: "replicaset-controller",
						},
						DeprecatedCount: 6,
						Note:            "(combined from similar events): Error creating: pods \"ws2-workspace-75dc7c5469-5trff\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server",
						Reason:          "FailedCreate",
						Type:            "Warning",
					},
					{
						DeprecatedFirstTimestamp: metav1.Date(2024, 5, 26, 15, 43, 16, 0, time.UTC),
						DeprecatedLastTimestamp:  metav1.Date(2024, 5, 26, 15, 43, 16, 0, time.UTC),
						DeprecatedSource: corev1.EventSource{
							Component: "replicaset-controller",
						},
						DeprecatedCount: 1,
						Note:            "Error creating: pods \"ws2-workspace-75dc7c5469-cxh42\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server",
						Reason:          "FailedCreate",
						Type:            "Warning",
					},
				},
			},
			want: []eventsv1.Event{
				{
					DeprecatedFirstTimestamp: metav1.Date(2024, 5, 26, 15, 43, 16, 0, time.UTC),
					DeprecatedLastTimestamp:  metav1.Date(2024, 5, 26, 15, 43, 16, 0, time.UTC),
					DeprecatedSource: corev1.EventSource{
						Component: "replicaset-controller",
					},
					DeprecatedCount: 1,
					Note:            "Error creating: pods \"ws2-workspace-75dc7c5469-cxh42\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server",
					Reason:          "FailedCreate",
					Type:            "Warning",
				},
				{
					DeprecatedFirstTimestamp: metav1.Date(2024, 5, 26, 15, 43, 16, 0, time.UTC),
					DeprecatedLastTimestamp:  metav1.Date(2024, 5, 26, 15, 44, 35, 0, time.UTC),
					DeprecatedSource: corev1.EventSource{
						Component: "replicaset-controller",
					},
					DeprecatedCount: 6,
					Note:            "(combined from similar events): Error creating: pods \"ws2-workspace-75dc7c5469-5trff\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server",
					Reason:          "FailedCreate",
					Type:            "Warning",
				},
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			sorted(tt.args.events)
			if !slices.EqualFunc(tt.args.events, tt.want, func(a, b eventsv1.Event) bool { return reflect.DeepEqual(a, b) }) {
				t.Errorf("sorted() diff %s", cmp.Diff(tt.args.events, tt.want))
			}
		})
	}
}
