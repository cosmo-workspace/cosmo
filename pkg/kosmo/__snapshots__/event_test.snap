['event Eventf emit events 1']
SnapShot = 'Normal TEST1 Instance: key1=value1 map[cosmo-workspace.github.io/instance:inst1 cosmo-workspace.github.io/user:tom]'

['event Eventf emit events 2']
SnapShot = 'Normal TEST2 ClusterInstance: key1=value1 map[cosmo-workspace.github.io/instance:cinst1]'

['event Eventf emit events 3']
SnapShot = 'Warning TEST3 Addon: key1=value1 map[cosmo-workspace.github.io/instance:cinst2 cosmo-workspace.github.io/user:user1]'

['event Eventf emit events 4']
SnapShot = 'Warning TEST4 User: key1=value1 map[cosmo-workspace.github.io/user:user2]'

['event when list user events should return a array of the events in user namespace and the user annotated events in default namespace 1']
SnapShot = """
Event Time                                                Type      Reason                Regarding                                                 Reporting Controller                Note
time.Date(2024, time.May, 27, 0, 41, 54, 0, time.Local)   Normal    Synced                User/xxx                                                  cosmo-user-controller               Successfully reconciled. Namespace cosmo-user-xxx is created
time.Date(2024, time.May, 27, 0, 41, 54, 0, time.Local)   Normal    PasswordInitialized   User/xxx                                                  cosmo-user-controller               Successfully reset password secret
time.Date(2024, time.May, 27, 0, 41, 54, 0, time.Local)   Normal    AddonSynced           User/xxx                                                  cosmo-user-controller               Addon cosmo-username-headers is created
time.Date(2024, time.May, 27, 0, 41, 54, 0, time.Local)   Normal    AddonSynced           User/xxx                                                  cosmo-user-controller               Addon resource-limitter is created
time.Date(2024, time.May, 27, 0, 41, 54, 0, time.Local)   Warning   AddonFailed           User/xxx                                                  cosmo-user-controller               Failed to sync addon: failed to create or update addon gryffindor-serviceaccount :failed to set controller reference: cluster-scoped resource must not have a namespace-scoped owner, owner's namespace cosmo-user-xxx
time.Date(2024, time.May, 27, 0, 41, 54, 0, time.Local)   Normal    AddonSynced           User/xxx                                                  cosmo-user-controller               Addon gryffindor-serviceaccount is created
time.Date(2024, time.May, 27, 0, 41, 54, 0, time.Local)   Normal    Synced                Instance/useraddon-cosmo-username-headers                 cosmo-instance-controller           Middleware cosmo-username-headers is created
time.Date(2024, time.May, 27, 0, 41, 54, 0, time.Local)   Normal    Synced                Instance/useraddon-resource-limitter                      cosmo-instance-controller           ResourceQuota quota is created
time.Date(2024, time.May, 27, 0, 41, 54, 0, time.Local)   Normal    Synced                ClusterInstance/useraddon-xxx-gryffindor-serviceaccount   cosmo-cluster-instance-controller   ServiceAccount iamserviceaccount is created
time.Date(2024, time.May, 27, 0, 41, 54, 0, time.Local)   Normal    Synced                ClusterInstance/useraddon-xxx-gryffindor-serviceaccount   cosmo-cluster-instance-controller   ClusterRoleBinding xxx-view is created
time.Date(2024, time.May, 27, 0, 43, 13, 0, time.Local)   Normal    Synced                Workspace/ws2                                             cosmo-workspace-controller          Successfully reconciled. Instance ws2 is created
time.Date(2024, time.May, 27, 0, 43, 13, 0, time.Local)   Normal    Synced                Workspace/ws2                                             cosmo-workspace-controller          Successfully reconciled. Traefik ingress route ws2 is created
time.Date(2024, time.May, 27, 0, 43, 13, 0, time.Local)   Normal    Synced                Instance/ws2                                              cosmo-instance-controller           Service ws2-workspace is created
time.Date(2024, time.May, 27, 0, 43, 13, 0, time.Local)   Normal    Synced                Instance/ws2                                              cosmo-instance-controller           Deployment ws2-workspace is created
time.Date(2024, time.May, 27, 0, 43, 13, 0, time.Local)   Normal    ScalingReplicaSet     Deployment/ws2-workspace                                  deployment-controller               Scaled up replica set ws2-workspace-75dc7c5469 to 1
time.Date(2024, time.May, 27, 0, 43, 13, 0, time.Local)   Warning   FailedCreate          ReplicaSet/ws2-workspace-75dc7c5469                       replicaset-controller               Error creating: pods \"ws2-workspace-75dc7c5469-sf424\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server
time.Date(2024, time.May, 27, 0, 43, 13, 0, time.Local)   Warning   FailedCreate          ReplicaSet/ws2-workspace-75dc7c5469                       replicaset-controller               Error creating: pods \"ws2-workspace-75dc7c5469-z42z2\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server
time.Date(2024, time.May, 27, 0, 43, 13, 0, time.Local)   Warning   FailedCreate          ReplicaSet/ws2-workspace-75dc7c5469                       replicaset-controller               Error creating: pods \"ws2-workspace-75dc7c5469-xr8sx\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server
time.Date(2024, time.May, 27, 0, 43, 13, 0, time.Local)   Warning   FailedCreate          ReplicaSet/ws2-workspace-75dc7c5469                       replicaset-controller               Error creating: pods \"ws2-workspace-75dc7c5469-vpqhz\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server
time.Date(2024, time.May, 27, 0, 43, 14, 0, time.Local)   Warning   FailedCreate          ReplicaSet/ws2-workspace-75dc7c5469                       replicaset-controller               Error creating: pods \"ws2-workspace-75dc7c5469-wmj97\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server
time.Date(2024, time.May, 27, 0, 43, 14, 0, time.Local)   Warning   FailedCreate          ReplicaSet/ws2-workspace-75dc7c5469                       replicaset-controller               Error creating: pods \"ws2-workspace-75dc7c5469-qb54v\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server
time.Date(2024, time.May, 27, 0, 43, 14, 0, time.Local)   Warning   FailedCreate          ReplicaSet/ws2-workspace-75dc7c5469                       replicaset-controller               Error creating: pods \"ws2-workspace-75dc7c5469-87k4l\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server
time.Date(2024, time.May, 27, 0, 43, 14, 0, time.Local)   Warning   FailedCreate          ReplicaSet/ws2-workspace-75dc7c5469                       replicaset-controller               Error creating: pods \"ws2-workspace-75dc7c5469-mdmg6\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server
time.Date(2024, time.May, 27, 0, 43, 15, 0, time.Local)   Warning   FailedCreate          ReplicaSet/ws2-workspace-75dc7c5469                       replicaset-controller               Error creating: pods \"ws2-workspace-75dc7c5469-cxh42\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server
time.Date(2024, time.May, 27, 0, 44, 35, 0, time.Local)   Warning   FailedCreate          ReplicaSet/ws2-workspace-75dc7c5469                                                           (combined from similar events): Error creating: pods \"ws2-workspace-75dc7c5469-5trff\" is forbidden: failed quota: quota: must specify limits.cpu for: code-server
"""
